File: android/app/src/main/java/com/nextpage/data/remote/addons/AddonRegistry.kt
package com.nextpage.data.remote.addons

import com.nextpage.data.local.dao.AddonDao
import com.nextpage.data.local.entity.AddonEntity
import com.nextpage.data.remote.catalog.CatalogProvider
import com.nextpage.data.remote.catalog.isHttpSuccess
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

/**
 * Row shape for [catalogProvidersWithAddons]; mirrors desktop `InstalledAddonRow`.
 */
data class InstalledAddonRow(
    val id: String,
    val url: String,
    val manifest: AddonManifest,
    val enabled: Boolean,
    val addedAt: Long
)

/**
 * Minimal install-by-URL seam contract (settings ViewModel + rebuild callers).
 *
 * Additive addon-deeplink-v1 surface: [fetchManifest] exposes fetch+validate
 * (preview, no persistence) and [installManifest] persists an already-fetched
 * manifest - the confirm-install split. `install()` keeps its atomic
 * fetch+validate+persist behavior.
 */
interface AddonRegistryLike {
    suspend fun listInstalled(): List<InstalledAddonRow>
    suspend fun install(url: String): AddonManifest
    suspend fun fetchManifest(url: String): AddonManifest
    suspend fun installManifest(url: String, manifest: AddonManifest): AddonManifest
    suspend fun setEnabled(id: String, enabled: Boolean)
    suspend fun uninstall(id: String)
}

/**
 * Install-by-URL service over the [AddonDao] registry (Android mirror of desktop
 * `AddonRegistry.ts`). install(url): HTTPS check → platform transport →
 * ManifestValidator → sha256 addonId → UPSERT preserving enabled. Enable/
 * disable/uninstall mutate rows; the composite's existence-check cache guard
 * (design A1) keeps uninstalled or disabled addon sources from ever serving
 * cached entries.
 */
class AddonRegistry(
    private val dao: AddonDao,
    private val transport: AddonHttpTransport,
    private val now: () -> Long = { System.currentTimeMillis() }
) : AddonRegistryLike {

    private val changeListeners = mutableListOf<(Int) -> Unit>()
    private var mutationVersion = 0

    /** Live-composite hook: fires after every registry mutation. */
    fun addOnChangedListener(listener: (Int) -> Unit) {
        changeListeners.add(listener)
    }

    private fun notifyChanged() {
        mutationVersion += 1
        for (listener in changeListeners.toList()) listener(mutationVersion)
    }

    /** install(url): HTTPS check → fetch → validate → addonId → upsert preserving enabled. */
    override suspend fun install(url: String): AddonManifest {
        val manifest = fetchManifest(url)
        return installManifest(url, manifest)
    }

    /**
     * Fetch + validate only (no persistence) - the deep-link preview step.
     * HTTPS is enforced here so a preview can never trigger non-https I/O.
     */
    override suspend fun fetchManifest(url: String): AddonManifest {
        ManifestValidator.assertHttpsInstallUrl(url)
        val fetched = transport.fetch(url)
        if (!fetched.status.isHttpSuccess()) {
            throw AddonFetchException(AddonFetchErrorCode.NETWORK, "addon fetch status ${fetched.status}")
        }
        return ManifestValidator.validate(fetched.body, fetched.contentType)
    }

    /**
     * Persist an already-fetched manifest (sha256 addonId → upsert preserving
     * enabled/addedAt → notifyChanged). No fetch; built-in id collision guard.
     */
    override suspend fun installManifest(url: String, manifest: AddonManifest): AddonManifest {
        if (
            manifest.id in BUILTIN_SOURCE_NAMES ||
            manifest.id.contains(':') ||
            manifest.id.startsWith("builtin")
        ) {
            throw AddonFetchException(
                AddonFetchErrorCode.INVALID_MANIFEST,
                "addon id must not collide with built-in source ids: ${manifest.id}"
            )
        }
        val addonId = AddonId.fromUrl(url)
        val existing = dao.getById(addonId)
        dao.upsert(
            AddonEntity(
                id = addonId,
                url = url,
                manifestJson = JSONObject(
                    mapOf(
                        "id" to manifest.id,
                        "name" to manifest.name,
                        "version" to manifest.version,
                        "catalogs" to org.json.JSONArray().apply {
                            manifest.catalogs.forEach {
                                put(
                                    org.json.JSONObject()
                                        .put("type", it.type)
                                        .put("id", it.id)
                                        .put("name", it.name)
                                )
                            }
                        },
                        "resources" to org.json.JSONArray().apply { manifest.resources.forEach { put(it) } }
                    )
                ).apply {
                    manifest.searchUrl?.let { put("searchUrl", it) }
                    manifest.detailsUrl?.let { put("detailsUrl", it) }
                }.toString(),
                enabled = existing?.enabled ?: true,
                addedAt = existing?.addedAt ?: now()
            )
        )
        notifyChanged()
        return manifest
    }

    /** Rows in install order (addedAt, then id tiebreak) with parsed manifests. */
    override suspend fun listInstalled(): List<InstalledAddonRow> =
        withContext(Dispatchers.IO) { dao.getAll() }
            .mapNotNull { row ->
                runCatching { AddonManifestJson.decode(row.manifestJson) }.getOrNull()?.let { manifest ->
                    InstalledAddonRow(row.id, row.url, manifest, row.enabled, row.addedAt)
                }
            }
            .sortedWith(compareBy({ it.addedAt }, { it.id }))

    override suspend fun setEnabled(id: String, enabled: Boolean) {
        try {
            dao.setEnabled(id, enabled)
        } finally {
            notifyChanged()
        }
    }

    override suspend fun uninstall(id: String) {
        try {
            dao.delete(id)
        } finally {
            notifyChanged()
        }
    }

    private companion object {
        /** Closed set of built-in source names addon manifests must not claim. */
        val BUILTIN_SOURCE_NAMES = setOf("gutendex", "openlibrary")
    }
}

/** JSON codecs for AddonManifest (org.json, mirroring ManifestValidator's parser). */
internal object AddonManifestJson {
    fun decode(json: String): AddonManifest = ManifestValidator.validate(
        json.toByteArray(Charsets.UTF_8),
        "application/json"
    )

    fun encode(manifest: AddonManifest): String = JSONObject()
        .put("id", manifest.id)
        .put("name", manifest.name)
        .put("version", manifest.version)
        .put(
            "catalogs",
            org.json.JSONArray().apply {
                manifest.catalogs.forEach {
                    put(
                        org.json.JSONObject()
                            .put("type", it.type)
                            .put("id", it.id)
                            .put("name", it.name)
                    )
                }
            }
        )
        .put("resources", org.json.JSONArray().apply { manifest.resources.forEach { put(it) } })
        .apply {
            manifest.searchUrl?.let { put("searchUrl", it) }
            manifest.detailsUrl?.let { put("detailsUrl", it) }
        }
        .toString()
}

/**
 * Ordered provider list: built-ins first, curated, then enabled addons in
 * install order (disabled rows excluded). Mirrors desktop `defaultCatalogProviders`.
 */
fun catalogProvidersWithAddons(
    builtIns: List<CatalogProvider>,
    curated: CatalogProvider,
    installedAddons: List<InstalledAddonRow>,
    addonTransport: AddonHttpTransport? = null
): List<CatalogProvider> =
    builtIns + curated + installedAddons
        .filter { it.enabled }
        .map { AddonCatalogProvider(it.manifest, it.id, addonTransport) }

File: desktop/src/lib/shared/services/addons/AddonRegistry.ts
/**
 * AddonRegistry — install-by-URL service over the desktop registry store.
 * install(url): HTTPS check → platform fetch → validateManifest → addonId →
 * UPSERT preserving enabled. enable/disable/uninstall mutate rows; the
 * composite's existence-check cache guard (design A1) keeps uninstalled or
 * disabled addon sources from ever serving cached entries.
 */
import { invoke } from '@tauri-apps/api/core';

import { addonIdFromUrl } from './addonId';
import {
  AddonFetchError,
  AddonFetchErrorCode,
  assertHttpsInstallUrl,
  validateManifest,
  type AddonManifest,
} from './validateManifest';

/** Closed set of built-in source names addon manifests must not claim. */
const BUILTIN_SOURCE_NAMES = new Set(['gutendex', 'openlibrary']);

export { AddonFetchErrorCode, AddonFetchError };
export type { AddonManifest };

export interface InstalledAddonRow {
  id: string;
  url: string;
  manifest: AddonManifest;
  enabled: boolean;
  addedAt: number;
}

/** Raw persistence row; mirrors Rust `InstalledAddonDto` / Room `AddonEntity`. */
interface RegistryStoreRow {
  id: string;
  url: string;
  manifestJson: string;
  enabled: boolean;
  addedAt: number;
}

export interface AddonRegistryStore {
  listInstalled(): Promise<RegistryStoreRow[]>;
  upsert(row: RegistryStoreRow): Promise<void>;
  setEnabled(id: string, enabled: boolean): Promise<void>;
  uninstall(id: string): Promise<void>;
}

/** Default store: the thin Rust CRUD commands (desktop only). */
export class TauriAddonRegistryStore implements AddonRegistryStore {
  async listInstalled(): Promise<RegistryStoreRow[]> {
    return invoke<RegistryStoreRow[]>('listInstalledAddons');
  }

  async upsert(row: RegistryStoreRow): Promise<void> {
    await invoke('upsertInstalledAddon', { addon: row });
  }

  async setEnabled(id: string, enabled: boolean): Promise<void> {
    await invoke('setAddonEnabled', { id, enabled });
  }

  async uninstall(id: string): Promise<void> {
    await invoke('deleteInstalledAddon', { id });
  }
}

export interface AddonFetchResult {
  status: number;
  contentType: string | null;
  body: Uint8Array;
}

/** Platform network layer seam (desktop: the Rust fetchAddonResource command). */
export type AddonTransport = (url: string) => Promise<AddonFetchResult>;

/**
 * Platform network layer seam (desktop: the Rust fetchAddonResource command
 * - addon hosts are never fetched from the webview). Used by the install
 * flow and AddonCatalogProvider; tests inject fakes via the constructors.
 */
export function defaultAddonTransport(): AddonTransport {
  return async (url) => {
    const res = await invoke<{ status: number; contentType: string | null; body: number[] }>(
      'fetchAddonResource',
      { url },
    );
    return { status: res.status, contentType: res.contentType, body: new Uint8Array(res.body) };
  };
}

export class AddonRegistry {
  private readonly store: AddonRegistryStore;
  private readonly transport: AddonTransport;
  private readonly now: () => number;
  private readonly listeners = new Set<(version: number) => void>();
  private mutationVersion = 0;

  constructor(options: {
    store?: AddonRegistryStore;
    transport?: AddonTransport;
    now?: () => number;
  } = {}) {
    this.store = options.store ?? new TauriAddonRegistryStore();
    this.transport = options.transport ?? defaultAddonTransport();
    this.now = options.now ?? (() => Date.now());
  }

  async install(url: string): Promise<AddonManifest> {
    assertHttpsInstallUrl(url);
    const fetched = await this.transport(url);
    if (fetched.status < 200 || fetched.status >= 300) {
      throw new AddonFetchError(AddonFetchErrorCode.NETWORK, `addon fetch status ${fetched.status}`);
    }
    const manifest = validateManifest(fetched.body, fetched.contentType);
    if (
      BUILTIN_SOURCE_NAMES.has(manifest.id) ||
      manifest.id.includes(':') ||
      manifest.id.startsWith('builtin')
    ) {
      throw new AddonFetchError(
        AddonFetchErrorCode.INVALID_MANIFEST,
        `addon id must not collide with built-in source ids: ${manifest.id}`,
      );
    }
    const addonId = await addonIdFromUrl(url);
    const existing = (await this.store.listInstalled()).find((row) => row.id === addonId);
    await this.store.upsert({
      id: addonId,
      url,
      manifestJson: JSON.stringify(manifest),
      enabled: existing ? existing.enabled : true,
      addedAt: existing ? existing.addedAt : this.now(),
    });
    this.notifyChanged();
    return manifest;
  }

  async listInstalled(): Promise<InstalledAddonRow[]> {    const rows = await this.store.listInstalled();
    return rows
      .map((row) => {
        try {
          return { row, manifest: JSON.parse(row.manifestJson) as AddonManifest };
        } catch {
          return { row, manifest: null };
        }
      })
      .filter((entry): entry is { row: RegistryStoreRow; manifest: AddonManifest } => entry.manifest !== null)
      .map(({ row, manifest }) => ({
        id: row.id,
        url: row.url,
        manifest,
        enabled: row.enabled,
        addedAt: row.addedAt,
      }))
      .sort((a, b) => a.addedAt - b.addedAt || a.id.localeCompare(b.id));
  }

  async setEnabled(id: string, enabled: boolean): Promise<void> {
    try {
      await this.store.setEnabled(id, enabled);
    } finally {
      this.notifyChanged();
    }
  }

  async uninstall(id: string): Promise<void> {
    try {
      await this.store.uninstall(id);
    } finally {
      this.notifyChanged();
    }
  }

  /**
   * Live-composite hook: fires after every registry mutation (install,
   * enable/disable, uninstall) with a monotonically increasing version.
   * Returns an unsubscribe function.
   */
  onChanged(listener: (version: number) => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  private notifyChanged(): void {
    this.mutationVersion += 1;
    for (const listener of [...this.listeners]) {
      try {
        listener(this.mutationVersion);
      } catch {
        // listener failures must never break the mutation flow
      }
    }
  }
}
